src/ica_api.c

```diff
@@ -1200,27 +1200,29 @@ unsigned int ica_rsa_crt_key_check(ica_rsa_key_crt_t *rsa_key)
 		keyfmt = 0;
 
 	if (!keyfmt) {
-		/* swap p and q */
-		tmp_buf = calloc(1, rsa_key->key_length/2);
+		tmp_buf = calloc(1, rsa_key->key_length / 2);
 		if (!tmp_buf)
 			return ENOMEM;
-		memcpy(tmp_buf, rsa_key->p + 8, rsa_key->key_length/2);
-		memcpy(rsa_key->p + 8, rsa_key->q, rsa_key->key_length/2);
-		memcpy(rsa_key->q, tmp_buf, rsa_key->key_length/2);
 
-		/* swap dp and dq */
-		memcpy(tmp_buf, rsa_key->dp + 8, rsa_key->key_length/2);
-		memcpy(rsa_key->dp + 8, rsa_key->dq, rsa_key->key_length/2);
-		memcpy(rsa_key->dq, tmp_buf, rsa_key->key_length/2);
-
-		/* calculate new qInv */
-		bn_p = BN_new();
-		bn_q = BN_new();
-		bn_invq = BN_new();
+		bn_p = BN_secure_new();
+		bn_q = BN_secure_new();
+		bn_invq = BN_secure_new();
 		ctx = BN_CTX_new();
+		if (!bn_p || !bn_q || !bn_invq || !ctx)
+			return ENOMEM;
```

Comment:
You need to free the BIGNUMs and the CTX in this error case.
At least one if it is NULL, but the others might be non-NULL and thus must be freed.

Reply:
Yes, right. I changed the logic a bit so that we always run through the free / clear_free.